File: java/flight/flight-jdbc/src/test/java/org/apache/arrow/jdbc/DriverTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.arrow.jdbc;

import static org.junit.Assert.*;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.Types;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

import org.junit.Ignore;
import org.junit.Test;

import com.google.common.collect.ImmutableList;

/**
 * JDBC Driver unit tests.
 */
public class DriverTest {

  final Driver driver = new org.apache.arrow.jdbc.Driver();

  @Test
  public void acceptsValidUrl() throws SQLException {
    assertTrue(driver.acceptsURL("jdbc:arrow://localhost:50051"));
  }

  @Test
  public void rejectsInvalidUrl() throws SQLException {
    assertFalse(driver.acceptsURL("jdbc:mysql://localhost:50051"));
  }

  @Test
  public void rejectsNullUrl() throws SQLException {
    assertFalse(driver.acceptsURL(null));
  }

  /**
   * Note that this is a manual integration test that requires the Rust flight-server example to be running.
   */
  @Test
  @Ignore
  public void executeQuery() throws SQLException {
    try (Connection conn = DriverManager.getConnection("jdbc:arrow://localhost:50051", new Properties())) {
      try (Statement stmt = conn.createStatement()) {
        try (ResultSet rs = stmt.executeQuery("SELECT id FROM alltypes_plain")) {

          ResultSetMetaData md = rs.getMetaData();
          assertEquals(1, md.getColumnCount());
          assertEquals("c0", md.getColumnName(1));
          assertEquals(Types.INTEGER, md.getColumnType(1));

          List<Integer> ids = new ArrayList<>();
          while (rs.next()) {
            ids.add(rs.getInt(1));
          }
          assertEquals(ImmutableList.of(4, 5, 6, 7, 2, 3, 0, 1), ids);
        }
      }
    }
  }
}

File: cpp/src/arrow/engine/substrait/driver_test.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include <gtest/gtest.h>

#include "arrow-adbc/adbc.h"
#include "arrow/engine/substrait/driver.h"

// Self-contained version of the Handle
static inline void clean_up(AdbcDriver* ptr) { ptr->release(ptr, nullptr); }

// static inline void clean_up(AdbcDatabase* ptr) {
//   ptr->private_driver->DatabaseRelease(ptr, nullptr);
// }

// static inline void clean_up(AdbcConnection* ptr) {
//   ptr->private_driver->ConnectionRelease(ptr, nullptr);
// }

// static inline void clean_up(AdbcStatement* ptr) {
//   ptr->private_driver->StatementRelease(ptr, nullptr);
// }

// static inline void clean_up(AdbcError* ptr) {
//   if (ptr->release != nullptr) {
//     ptr->release(ptr);
//   }
// }

template <typename T>
class Handle {
 public:
  explicit Handle(T* value) : value_(value) {}

  ~Handle() { clean_up(value_); }

 private:
  T* value_;
};

TEST(AdbcDriver, AdbcDriverInit) {
  // Test the get/set option implementation in the base driver
  struct AdbcDriver driver;
  memset(&driver, 0, sizeof(driver));
  ASSERT_EQ(arrow::engine::AceroDriverInitFunc(ADBC_VERSION_1_1_0, &driver, nullptr),
            ADBC_STATUS_OK);
  Handle<AdbcDriver> driver_handle(&driver);
}
